File: livedatademo/src/main/java/com/jarvis/livedatademo/NameActivity.java
package com.jarvis.livedatademo;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProvider;

import android.os.Bundle;
import android.os.PersistableBundle;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.TextView;

import java.util.Random;

public class NameActivity extends AppCompatActivity {

    private TextView tvName;
    private Button change;
    private NameViewModel model;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_name);
        Log.i("NameActivity：", "onCreate()");

        tvName = findViewById(R.id.tv_name);
        change = findViewById(R.id.change);


        model = new ViewModelProvider(this).get(NameViewModel.class);
        final Observer<String> observer = s -> {
            Log.i("Jarvis:",s);
            tvName.setText(s);
        };
        model.getCurrentName().observe(this, observer);

        change.setOnClickListener(v -> {
            String name = "Hello" + model.getI();
            model.getCurrentName().setValue(name);
            model.getCurrentName().setValue(name + "1");


        });

    }


    @Override
    public void onSaveInstanceState(@NonNull Bundle outState, @NonNull PersistableBundle outPersistentState) {
        super.onSaveInstanceState(outState, outPersistentState);
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        Log.i("NameActivity：", "onDestroy()");
    }
}

File: livedatademo/src/main/java/com/jarvis/livedatademo/NameActivity.kt
package com.jarvis.livedatademo

import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import androidx.activity.viewModels
import androidx.databinding.DataBindingUtil
import androidx.lifecycle.Observer
import com.jarvis.livedatademo.databinding.ActivityName2Binding

class NameActivity : AppCompatActivity() {

    private lateinit var bind: ActivityName2Binding
    private val model: NameViewModel by viewModels()
    private lateinit var userInfo1: UserInfo

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        userInfo1 = UserInfo("11", 1)
        bind = DataBindingUtil.setContentView(this, R.layout.activity_name2)

        //创建观察者
        val nameObserver = Observer<String> { newName ->
            //更新UI
            bind.desc.text = newName
        }
        //订阅事件，同时传入 LifecycleOwner 和观察者
        model.currentName.observe(this, nameObserver)

        model.aa.observe(this) { userInfo ->
            userInfo1 = userInfo
            bind.desc.text = userInfo.name + userInfo.age
        }

        bind.btnName.setOnClickListener {
            model.useLocal.value = !(model.useLocal.value ?: true)
        }
        bind.changeUserName.setOnClickListener {
            model.userInfoNet.value = UserInfo("网络", 1 + userInfo1.age)
            model.userInfoLocal.value = UserInfo("本地", 1 + userInfo1.age)
        }

    }
}